### ISSUE TITLE
TaskVine occasionally doesn't run a task(?)

### ISSUE THREAD

**Issue by benclifford** [Parsl/parsl#3026]
**Describe the bug**

Here's something that's come up in some work @khk-globus has done rearranging our build system, so probably caused by shaking the race-condition machine -- @dthain @tphung3 

In one test run on a branch:

https://github.com/Parsl/parsl/actions/runs/7549882832/job/20555983849 with a zipfile of all the working files here: https://github.com/Parsl/parsl/actions/runs/7549882832/artifacts/1174063468

```
DEBUG    parsl.executors.taskvine.executor:executor.py:638 Updating Future for Parsl Task: 36.                                Task 36 has result_received set to False
...
parsl.executors.taskvine.errors.TaskVineTaskFailure: ('taskvine result: unable to process TaskVine system failure', None)
```

and in the working files collection, the relevant task (executor task 36 = taskvine task 37) does not have a result:

```
./000/TaskVineExecutor/function_data/0035/argument
./000/TaskVineExecutor/function_data/0035/function
./000/TaskVineExecutor/function_data/0035/map
./000/TaskVineExecutor/function_data/0035/result
./000/TaskVineExecutor/function_data/0036
./000/TaskVineExecutor/function_data/0036/argument
./000/TaskVineExecutor/function_data/0036/function
./000/TaskVineExecutor/function_data/0036/map
```

This happens:

```
1705460394612239 10624 TASK 37 READY generate FIRST_RESOURCES 1 {}
1705460395614276 10624 TASK 37 RUNNING worker-670a95fec2b0cbe6d1410e2aa269dce6  FIRST_RESOURCES {"time_commit_start":[1705460395.61355,"s"],"time_commit_end":[1705460395.614212,"s"],"time_input_mgr":[0.000662,"s"],"size_input_mgr":[0.008025169372558594,"MB"],"memory":[15991,"MB"],"disk":[23399,"MB"],"gpus":[0,"gpus"],"cores":[2,"cores"]}
1705460395622025 10624 TASK 37 WAITING_RETRIEVAL worker-670a95fec2b0cbe6d1410e2aa269dce6 
1705460395622123 10624 TASK 37 READY generate FIRST_RESOURCES 2 {}
1705460400448987 10624 TASK 37 RUNNING worker-670a95fec2b0cbe6d1410e2aa269dce6  FIRST_RESOURCES {"time_commit_start":[1705460400.448389,"s"],"time_commit_end":[1705460400.44894,"s"],"time_input_mgr":[0.0005509999999999999,"s"],"size_input_mgr":[0.008025169372558594,"MB"],"memory":[15991,"MB"],"disk":[23398,"MB"],"gpus":[0,"gpus"],"cores":[2,"cores"]}
1705460400449272 10624 TASK 37 WAITING_RETRIEVAL worker-670a95fec2b0cbe6d1410e2aa269dce6 
1705460400449394 10624 TASK 37 READY generate FIRST_RESOURCES 3 {}
1705460400449444 10624 TASK 37 RETRIEVED MAX_RETRIES  -1  {} {}
1705460400449464 10624 TASK 37 DONE MAX_RETRIES  -1 
```

I don't know where to look further than this to understand what is happening to the task inside taskvine.

=============
In the second failure, https://github.com/Parsl/parsl/actions/runs/7549882832/job/20554640838 which does not have any downloadable log collection, I see a task failing like this, which looks similar:

DEBUG    parsl.executors.taskvine.executor:executor.py:638 Updating Future for Parsl Task: 189.                                Task 189 has result_received set to False

**Comment by benclifford** [comment]
this is a fork of current master -  the taskvine executor code shouldn't be any different, and the cctools version used there is https://github.com/cooperative-computing-lab/cctools/releases/download/release/7.7.2/cctools-7.7.2-x86_64-ubuntu20.04.tar.gz

**Comment by dthain** [comment]
That sequence of events in the transaction log indicates that the task attempted to run twice.  After the second attempt, it exhausted the maximum number of retries and was returned to the caller.

@colinthomas-z80 please take a quick look at this (I think the TV debug log is the place to start) and figure out why it is failing.

**Comment by colinthomas-z80** [comment]
The task was forsaken the first time around. This also happened to task 22 but the second time it succeeded.

It looks like the problem is in the interaction between task requested disk space, and available disk at the worker. By default a task declaration requests the whole worker disk. Each time the worker computes its disk space it is marginally smaller because of the inputs. If a task goes through the resource setup process in between these events it will request more disk than the worker has available and the task will be forsaken. 

We will have to make an issue for this and think of a proper fix. You could work around this by manually declaring a disk space for the tasks such as `t.set_disk(1000)`

**Comment by benclifford** [comment]
that sounds quite similar to what @dthain hypothesised was happening with Work Queue here: https://github.com/cooperative-computing-lab/cctools/issues/3593



In the parsl case, maybe this could be changing:

```            if task.disk is not None:
                t.set_disk(task.disk)
```

to 

```
            if task.disk is not None:
                t.set_disk(0)
```

?

**Comment by colinthomas-z80** [comment]
Yes that should do it. I'm not sure what the other issue is about just yet.
